listenbrainz/webserver/views/feedback_api.py

```diff
@@ -0,0 +1,125 @@
+        log_raise_400("JSON document must contain recording_msid and "
+                      "score top level keys", data)
+
+    if 'recording_msid' in data and 'score' in data and len(data) > 2:
+        log_raise_400("JSON document may only contain recording_msid and "
+                      "score top level keys", data)
+
+    try:
+        feedback = Feedback(user_id=user["id"], recording_msid=data["recording_msid"], score=data["score"])
+    except ValidationError as e:
+        log_raise_400("Invalid JSON document submitted: %s" % str(e).replace("\n ", ":").replace("\n", " "),
+                      data)  # str.replace() to tidy up the error message
+
+    try:
+        if feedback.score == 0:
+            db_feedback.delete(feedback)
+        else:
+            db_feedback.insert(feedback)
+    except APIServiceUnavailable as e:
+        raise
+    except Exception as e:
+        raise APIInternalServerError("Something went wrong. Please try again.")
+
+    return jsonify({'status': 'ok'})
+
+
+@feedback_api_bp.route("/user/<user_name>/get-feedback", methods=["GET"])
+@crossdomain()
+@ratelimit()
+def get_feedback_for_user(user_name):
+    """
+    Get feedback given by user ``user_name``. The format for the JSON returned is defined in our :ref:`feedback-json-doc`.
+
+    If the optional argument ``score`` is not given, this endpoint will return all the feedback submitted by the user.
+    Otherwise filters the feedback to be returned by score.
+
+    :param score: If 1 then returns the loved recordings, if -1 returns hated recordings.
+    :statuscode 200: Yay, you have data!
+    :resheader Content-Type: *application/json*
+    """
+
+    score = _parse_int_arg('score')
+
+    user = db_user.get_by_mb_id(user_name)
+    if user is None:
+        raise APINotFound("Cannot find user: %s" % user_name)
+
+    if score is None:
+        feedback = db_feedback.get_feedback_by_user_id(user["id"])
+    else:
+        if score in [-1, 1]:
+            feedback = db_feedback.get_feedback_by_user_id_and_score(user["id"], score)
+        else:
+            log_raise_400("Score can have a value of 1 or -1.", request.args)
+
+    for i, fb in enumerate(feedback):
+        fb.user_id = user_name
+        feedback[i] = fb.__dict__
+
+    return jsonify(feedback)
+
+
+@feedback_api_bp.route("/recording/<recording_msid>/get-feedback", methods=["GET"])
+@crossdomain()
+@ratelimit()
+def get_feedback_for_recording(recording_msid):
+    """
+    Get feedback for recording with given ``recording_msid``. The format for the JSON returned
+    is defined in our :ref:`feedback-json-doc`.
+
+    :statuscode 200: Yay, you have data!
+    :resheader Content-Type: *application/json*
+    """
+
+    if not is_valid_uuid(recording_msid):
+        log_raise_400("%s MSID format invalid." % recording_msid)
+
+    feedback = db_feedback.get_feedback_by_recording_msid(recording_msid)
+
+    for i, fb in enumerate(feedback):
```

Comment:
This is a bad DB access pattern. If a recording has 1000 rows of feedback by 1000 different users, you'll execute 1,000 queries. You should make a list of users and then fetch them all using an SELECT ... FROM table WHERE user_name in ([list]).

Reply:
Oops. Improved.

Reply:
Oops. Updated. Thanks.